### PR TITLE
feat(balancer) prevent cascading failures

### LINKED ISSUE #25: Balancer: prevent cascading overload failures
when implementing the ability to fail a node, precaution should be made to prevent cascading failures in case of an overload (a failing node would increase the load on the remaining nodes, causing them to fail as well)

There is a lot of advanced stuff on "cascading failures", but since the balancer lives in a separate module, it has little information to act upon. As such it pretty much can only handle overload scenario's.

Initial thought on implementation was setting a limit on the failed "weight" in the balancer, beyond which fall-backs are skipped, and instead an error is returned. 

Consider a balancer with 10 nodes, each weight 10. So overall weight = 100. Setting the limit to 70%.
when 1 node is marked as unavailable, its traffic is rerouted to the other nodes. This causes a second node to fail, and then a third. When the 3rd node failed the failed "weight" is 30, and the available "weight" is 70, so the limit of 70% is now hit. From this point forward the balancer no longer falls through when it hits a failed node, but immediately returns an error.
This will now return errors on 30% of the traffic, but will prevent further cascading of the overload.

thoughts?

### PR DESCRIPTION
**Pull request by Tieske**
Implements a health measure. The relative "weight" in the balancer being disabled.

The new property `healthThreshold` can be used to set the minimum amount of healthy weight required (in %). When the actual percentage gets under the threshold, the balancer will return errors.

The default value for `healthThreshold == 0` hence default behaviour does not change. It is a breaking change though because the error message returned will be different, and the callback got a new event "health" whenever there is an update in health status.

fixes #25 

EDIT: rewrote the description since the initial description only covered the 1st commit

### CODE REVIEW

**Comment by Tieske** [comment]
@hishamhm there are some test failures, but they are unrelated. Some test dns records were deleted (all SRV related), I'm working on fixing those.

**Comment by Tieske** [comment]
Note that the last commit fixes linter errors, most of which existed before this pr.

**Comment by Tieske** [comment]
@hishamhm after discussion I tried to find slightly more intelligent back-off mechanisms. Problem is that most of those require application knowledge/client side behaviour.

To improve the following could be implemented;

1. apply a global retry-limit on a balancer (max 10 retries per second, or 200 per minute, etc). This would always be in effect.
2. Drop relative traffic (so 20% unhealthy means randomly dropping 20% of traffic) and stop handling retries all together. Applicable from a first threshold, eg. health drops below 70%.
3. Drop all traffic to stop the cascade. Applicable from a second threshold, eg. health drops below 50%.

This PR implements health metrics, and option 3 above. The other ones could be added later, since the third one is the most important (and most protective).

File: src/resty/dns/balancer/base.lua [hunk 1]
```diff
@@ -1076,9 +1077,14 @@ end
...
--- Call this either as `setPeerStatus(available, address)`, `setPeerStatus(available, handle)``, or as `setPeerStatus(available, ip, port, <hostname>)`.
+-- Call this either as:
+--
+-- - `setPeerStatus(available, address)`,
+-- - `setPeerStatus(available, handle)`, or as
+-- - `setPeerStatus(available, ip, port, hostname)`
+--
 -- Using the `address` or `handle` is preferred since it is guaranteed to match. By ip/port/name
--- might fail if there are too many DNS levels.
+-- might fail if there are to many DNS levels.
```
Comment (hishamhm) [inline src/resty/dns/balancer/base.lua]: this is a typo, the original was correct

Comment (Tieske) [reply src/resty/dns/balancer/base.lua]: fixed

**Comment by Tieske** [comment]
@locao would you mind having a look at this one as well. This is the one @hishamhm mentioned as "controversial" because of the all-or-nothing approach when reaching the threshold.

But as stated, the default value for the threshold is to kick in when everything has failed, hence it does not change default behaviour. Only when explicitly setting the threshold.

**Pull request merged**